### PR TITLE
Blog App: validations and model specs

### PR DESCRIPTION
**Pull request by SAZZAD1Q2**
Add the following validations:
For the User model:
        Name must not be blank.
        PostsCounter must be an integer greater than or equal to zero.
    For the Post model:
        Title must not be blank.
        Title must not exceed 250 characters.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
+  has_many :likes
+
+  attribute :name, :string
+  attribute :bio, :text
+  attribute :post_counter, :integer, default: 0
+  attribute :photo, :string
+
+  def recent_posts
+    posts.order(created_at: :desc).limit(3)
   end
```
Comment (nahnahdev) [inline app/models/user.rb:24]: - [ ] Great job so far @SAZZAD1Q2   There are a few changes I will suggest you implement and you'll be on your way to the next Milestone :wink: 
- [ ] Great job adding most of the tests, however, I noticed that per the project requirements, you were asked to `Add unit specs for all of your models' methods and validations.` But currently, this `recent_posts` method in your `User Model` hasn't been tested. So kindly include tests for that method as well :wink:

File: app/models/mynewpost_new.rb [hunk 2]
```diff
@@ -1,32 +1,28 @@
...
+  after_save :update_post_counter
 
+  def update_post_counter
+    author.update(post_counter: author.posts.count)
+  end
 
+  def recent_comments
+    comments.order(created_at: :desc).limit(5)
+  end
+end
```
Comment (nahnahdev) [inline app/models/mynewpost_new.rb:26]: - [ ] Currently, your `mynewpost_new_spec.rb` is also missing tests for these two methods in your `Post Model` so kindly include tests for them as well :wink:

File: app/models/mynewcomment_new.rb [hunk 3]
```diff
@@ -1,20 +1,20 @@
...
+  belongs_to :post, class_name: 'MynewpostNew'
+
+  attribute :text, :text
+
+  after_create :update_comment_counter
+  after_destroy :update_comment_counter
+
+  def update_comment_counter
+    post.update(comment_counter: post.comments.count)
   end
```
Comment (nahnahdev) [inline app/models/mynewcomment_new.rb:19]: - [ ] Kindly include tests for the custom method in your `Comment Model` as well. I have highlighted the method above. So kindly include tests for it :wink:

File: app/models/mynewlikes_new.rb [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
-      post.update(like_counter: post.likes.count)
-    end
+  belongs_to :user, class_name: 'User'
+  belongs_to :post, class_name: 'MynewpostNew'
+
+  after_save :update_like_counter
+
+  def update_like_counter
+    post.update(like_counter: post.likes.count)
+  end
```
Comment (nahnahdev) [inline app/models/mynewlikes_new.rb:11]: - [ ] Kindly include tests for the custom method in your `Like Model` as well. I have highlighted the method above. So kindly include tests for it :wink:

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -1,5 +1,22 @@
...
+
+  it 'is not valid without a name' do
+    user = User.new(post_counter: 0)
+    expect(user).not_to be_valid
+  end
+
+  it 'is not valid with a negative post counter' do
+    user = User.new(name: 'Alice', post_counter: -1)
+    expect(user).not_to be_valid
+  end
```
Comment (nahnahdev) [inline spec/models/user_spec.rb:21]: - [ ] After running the currently existing tests, most of them were failing. So kindly ensure that all your tests are passing and kindly include the missing tests as well :+1: 
- Failed Tests
image

**Review by nahnahdev — CHANGES_REQUESTED** [review]
## Project Status: CHANGES REQUIRED :recycle: 
Hi @SAZZAD1Q2    , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights :fire: 
- All required user validations added :heavy_check_mark: 
- All required posts validations added :heavy_check_mark: 
- Has unit tests for  `Post`, `Comment`, and `User` model :heavy_check_mark: 
- Model specs has **Validation** tests :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
- N/A

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/user_spec.rb (hunk 5 above)
Comment (SAZZAD1Q2) [reply spec/models/user_spec.rb:21]: Thank you for reviewing my project. But on my site I see no fail. Kindly will you inform me which command are you running to check my spec test case.

Comment (nahnahdev) [reply spec/models/user_spec.rb:21]: HY @SAZZAD1Q2 I used the `rspec spec` command

Comment (prolajumokeoni) [reply spec/models/user_spec.rb:21]: Screenshot_191
-Tests are passing welldone 💪🏿

**Review by prolajumokeoni — APPROVED** [review]
Hi @SAZZAD1Q2 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
## Highlights
 - Tests are pasing ✅ 
 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**